src/Interpreters/HashJoin.cpp

```diff
@@ -1365,19 +1385,29 @@ NO_INLINE IColumn::Filter joinRightColumns(
     constexpr JoinFeatures<KIND, STRICTNESS> join_features;
 
     size_t rows = added_columns.rows_to_add;
-    IColumn::Filter filter;
     if constexpr (need_filter)
-        filter = IColumn::Filter(rows, 0);
+        added_columns.filter = IColumn::Filter(rows, 0);
 
     Arena pool;
 
     if constexpr (join_features.need_replication)
         added_columns.offsets_to_replicate = std::make_unique<IColumn::Offsets>(rows);
 
     IColumn::Offset current_offset = 0;
-
-    for (size_t i = 0; i < rows; ++i)
+    size_t max_joined_block_rows = added_columns.max_joined_block_rows;
+    size_t i = 0;
+    for (; i < rows; ++i)
     {
+        if constexpr (join_features.need_replication)
```

Comment:
@vdimir you've explained what `need_replicaion` means at the time, but I've already forgotten and there is no comment. Let's add it next to the `need_replication` definition.

Reply:
Added comments and changed `resize_assume_reserved -> resize` https://github.com/ClickHouse/ClickHouse/pull/58289